lib/rubygems/core_ext/kernel_require.rb

```diff
@@ -59,6 +62,8 @@ def require(path) # :doc:
 
         Kernel.send(:gem, spec.name, Gem::Requirement.default_prerelease) unless
           resolved_path
+
+        next
```

Comment:
This change didn't seem to fix the issue for me. Doesn't this `next` just break out of the `RUBYGEMS_ACTIVATION_MONITOR.synchronize` block and continue as before?

Reply:
Yes, that's what it does!

Ooopps! Since I got the same error to reproduce with a spec and this change fixed it, I assumed it was also fixing your case. But your real world scenario seems to have even more complexity, I'll look more tomorrow.

Reply:
Thanks, that did indeed solve it, though should `find_unresolved_default_spec` be renamed to `find_default_spec`? 😄

Reply:
Good point, once again! I'll turn on paranoid mode and add `Gem.find_default_spec` and then deprecate `Gem.find_unresolved_default_spec` in a follow up PR.

Reply:
Can `next` be dropped?

Reply:
Which `next`? I think both fixes here are necessary.

Reply:
The tests still pass if I apply this diff to this branch:

```diff
diff --git a/lib/rubygems/core_ext/kernel_require.rb b/lib/rubygems/core_ext/kernel_require.rb
index 073966b696..ac652e5ed6 100644
--- a/lib/rubygems/core_ext/kernel_require.rb
+++ b/lib/rubygems/core_ext/kernel_require.rb
@@ -66,8 +66,6 @@ def require(path) # :doc:
 
         Kernel.send(:gem, name, Gem::Requirement.default_prerelease) unless
           resolved_path
-
-        next
       end
 
       # If there are no unresolved deps, then we can use just try
```

Reply:
Both tests added here fail for me without that line 🤔. In any case, even if tests passed, seems nice to save unnecessary work on first require of a default gem.

Reply:
@deivid-rodriguez I might be doing something wrong, but I'm seeing the same results with or without the `next` line (see below). I'm not sure whether `0 failures, 2 errors` means pass, but I think this is a test failure?

<details>

```
% bundle exec ruby test/rubygems/test_require.rb
Loaded suite test/rubygems/test_require
Started
http://example.com
.http://example.com
http://example.com
http://example.com
.http://example.com
http://example.com
http://example.com
http://example.com
.http://example.com
http://example.com
..http://example.com
http://example.com
http://example.com
http://example.com
.http://example.com
http://example.com
.http://example.com
..http://example.com
..http://example.com
http://example.com
http://example.com
http://example.com
E
==================================================================================================================================================================
Error: test_default_gem_required_circulary_with_unresolved_gems_depending_on_it(TestGemRequire): Gem::LoadError: can't activate net-http-0.1.1, already activated net-http-0.3.0
/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/specification.rb:2307:in `check_version_conflict'
/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/specification.rb:1446:in `activate'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:98:in `each'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:98:in `block in require'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:39:in `synchronize'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:39:in `require'
test/rubygems/test_require.rb:29:in `assert_require'
test/rubygems/test_require.rb:574:in `block in test_default_gem_required_circulary_with_unresolved_gems_depending_on_it'
/Users/stanhu/github/rubygems/test/rubygems/helper.rb:1253:in `silence_warnings'
test/rubygems/test_require.rb:573:in `test_default_gem_required_circulary_with_unresolved_gems_depending_on_it'
==================================================================================================================================================================
http://example.com
http://example.com
http://example.com
http://example.com
E
==================================================================================================================================================================
Error: test_default_gem_with_unresolved_gems_depending_on_it(TestGemRequire): Gem::LoadError: can't activate net-http-0.1.1, already activated net-http-0.3.0
/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/specification.rb:2307:in `check_version_conflict'
/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/specification.rb:1446:in `activate'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:98:in `each'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:98:in `block in require'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:39:in `synchronize'
<internal:/Users/stanhu/.asdf/installs/ruby/3.2.2/lib/ruby/site_ruby/3.2.0/rubygems/core_ext/kernel_require.rb>:39:in `require'
test/rubygems/test_require.rb:29:in `assert_require'
test/rubygems/test_require.rb:549:in `test_default_gem_with_unresolved_gems_depending_on_it'
==================================================================================================================================================================
.http://example.com
http://example.com
http://example.com
http://example.com
http://example.com
http://example.com
......http://example.com
..http://example.com
.http://example.com
http://example.com
.http://example.com
.http://example.com
.http://example.com
.http://example.com
.http://example.com
http://example.com
http://example.com
.http://example.com
http://example.com
http://example.com
.http://example.com
http://example.com
....http://example.com
..http://example.com
http://example.com
http://example.com
http://example.com
http://example.com
.
Finished in 2.115785 seconds.
------------------------------------------------------------------------------------------------------------------------------------------------------------------
37 tests, 115 assertions, 0 failures, 2 errors, 0 pendings, 0 omissions, 0 notifications
94.5946% passed
------------------------------------------------------------------------------------------------------------------------------------------------------------------
17.49 tests/s, 54.35 assertions/s
</details>

Reply:
It is! I'm not sure what's causing that but since current implementation passes CI (and locally for me) and fixes your realworld issue, I'd suggest to deal with issues in your testing environment as a separate issue?

Reply:
Actually, from looking at your test run output, I think the answer may be that tests are running against your system RubyGems, and you may not have manually applied the patch to that?